File: code/ch14/14.1.5_23.merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
   public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        PriorityQueue<ListNode> queue = new PriorityQueue<>(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                if (o1.val < o2.val) return -1;
                else if (o1.val == o2.val) return 0;
                else return 1;
            }
        });
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;
        for (ListNode node : lists) {
            if (node != null) queue.add(node);
        }
        while (!queue.isEmpty()) {
            p.next = queue.poll();
            p = p.next;
            if (p.next != null) queue.add(p.next);
        }
        return dummy.next;
    }

File: code/ch14/14.1.5_23.merge-k-sorted-lists.py
from typing import List
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        import heapq

        l = []

        for head in lists:
            if head:
                heapq.heappush(l, (head.val, head))

        dummy = ListNode(-1)
        cur = dummy

        while l:
            _, node = heapq.heappop(l)
            cur.next = node
            cur = cur.next
            node = node.next
            if node:
                heapq.heappush(l, (node.val, node))

        return dummy.next
